README.md
# Sensu Go Handler Event KeepAlive By TTL 
![Go Test](https://github.com/jefferson22alcantara/sensu-event-keepalive-ttl-handler/workflows/Go%20Test/badge.svg)
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/jefferson22alcantara/sensu-event-keepalive-ttl-handler)

The Sensu Go Handler Event KeepAlive By TTL is a [Sensu Event Handler][2] which manages
Sensu Events  Generede by ad hoc Checks that still alerting after TTL expired as message log  **`"level":"warning","msg":"check TTL expired"`** and **`"output": "Last check execution was xX seconds ago"`** .The purposes of handler is [Sensu][1] remove that events after number of occurrences trigged.


## Installation

Download the latest version of the sensu-event-keepalive-ttl-handler from [releases][3],
or create an executable script from this source.

From the local path of the sensu-event-keepalive-ttl-handler repository:
```
go build -o /usr/local/bin/sensu-event-keepalive-ttl-handler main.go
```

## Configuration

Example Sensu Go handler definition:

```yml
type: Handler
api_version: core/v2
metadata:
  name: event-keepalive-ttl-handler
  namespace: default
spec:
  type: pipe
  command: sensu-event-keepalive-ttl-handler -k APYKEY -h http://localhost:8080 -m 10
  env_vars:
  - API_JEY=""
  - SENSU_MAX_OCCURRENCE=10
  - SENSU_URL=http://localhost:8080
  timeout: 10
  runtime_assets:
  - jefferson22alcantara/sensu-event-keepalive-ttl-handler
  filters:
  - is_incident
```

Example Sensu Go check definition:

```yml
---
type: CheckConfig
api_version: core/v2
metadata:
  name: dummy-app-healthz
  namespace: default
  annotations:
    sensu.io/plugins/sensu-event-keepalive-ttl-handler/conf/enable: "true"
spec:
  command: check-http -u http://localhost:8080/healthz
  subscriptions:
  - dummy
  handlers:
  - event-keepalive-ttl-handler
  interval: 60
  publish: true
```


## Usage Examples

Help:
```


The Sensu Go Remove events as keepalive

Usage:
  sensu-event-keepalive-ttl-handler [flags]
  sensu-event-keepalive-ttl-handler [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -k, --apikey string        The Sensu Api Key , use default from API_JEY env var
  -h, --help                 help for sensu-event-keepalive-ttl-handler
  -m, --max_occurrence int   The Max event Occurence after ttl to remove event   , use default from SENSU_MAX_OCCURRENCE env var
  -u, --sensu_url string     The Sensu Api URL  , use default from SENSU_URL env var

Use "sensu-event-keepalive-ttl-handler [command] --help" for more information about a co

```

**Note:** Make sure to set the `"sensu.io/plugins/sensu-event-keepalive-ttl-handler/conf/enable": "true",` annotaition on checks. 

### Argument Annotations

All arguments for this handler are tunable on a per entity or check basis based on annotations.  The
annotations keyspace for this handler is `"sensu.io/plugins/sensu-event-keepalive-ttl-handler/conf/enable": "true",`. 

#### Examples

To change the team argument for a particular check, for that checks's metadata add the following:

```yml
type: CheckConfig
api_version: core/v2
metadata:
  annotations:
    sensu.io/plugins/sensu-event-keepalive-ttl-handler/conf/enable: "true"
[...]
```


### Asset creation

The easiest way to get this handler added to your Sensu environment, is to add it as an asset from Bonsai:

```sh
sensuctl asset add jefferson22alcantara/sensu-event-keepalive-ttl-handler
```

See `sensuctl asset --help` for details on how to specify version.

## Contributing

See https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md

[1]: https://github.com/sensu/sensu-go
[2]: https://docs.sensu.io/sensu-go/5.0/reference/handlers/#how-do-sensu-handlers-work
[3]: https://github.com/jefferson22alcantara/sensu-event-keepalive-ttl-handler/releases


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).


## [0.0.2] - 2021-03-06

### Added
- Initial release


main_test.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"regexp"
	"testing"

	"github.com/sensu-community/sensu-plugin-sdk/httpclient"
	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/sensu/sensu-go/types"
)

var server = httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
	if req.Method == http.MethodGet {
		event := corev2.FixtureEvent("localhost", "keepalive")
		_ = json.NewEncoder(w).Encode(event)
	}
}))

// Testannotations teste
func Testannotations(t *testing.T) (bool, error) {
	anotation := make(map[string]string)
	anotation["sensu.io/plugins/sensu-event-keepalive-ttl-handler/conf/enable"] = "true"
	event := &types.Event{
		ObjectMeta: corev2.ObjectMeta{Annotations: anotation},
	}
	if event.Check.Annotations != nil {
		for key, value := range event.Check.Annotations {
			if key == "sensu.io/plugins/sensu-event-keepalive-ttl-handler/conf/enable" {
				if value == "true" {
					return true, nil
				}

			}

		}

	}
	return false, nil
}

//Testparameter test
func Testparameter(t *testing.T) bool {
	event := &types.Event{
		Timestamp:            0,
		Entity:               &types.Entity{ObjectMeta: types.ObjectMeta{Name: "EntityName"}},
		Check:                &types.Check{ObjectMeta: types.ObjectMeta{Name: "CheckName"}, Output: "Last check execution was 30 seconds ago"},
		Metrics:              &corev2.Metrics{},
		ObjectMeta:           corev2.ObjectMeta{},
		ID:                   []byte{},
		Sequence:             0,
		XXX_NoUnkeyedLiteral: struct{}{},
		XXX_unrecognized:     []byte{},
		XXX_sizecache:        0,
	}
	plugin.MAXOCCURENCE = 60

	re, err := regexp.Compile(`\w+\s\w+\s\w+\s\w+\s\d+\s\w+\s\w+`)
	if err != nil {
		fmt.Println("Regex Compile Error ")
		return false
	}

	if re.MatchString(event.Check.Output) {
		fmt.Println("Event Check Output [PASS]", event.Check.Output)
	} else {
		fmt.Println("Event Check Output [FAIL] Event not Match with Regex ttl [nothing to do]", event.Check.Output)
		return false
	}

	if event.Check.Status == 1 {
		fmt.Println("Event Check Status [PASS]", event.Check.Status)
	} else {
		fmt.Println("Event Check Status [FAIL], Event not Match with Check Status 1 [nothing to do]", event.Check.Status)
		return false
	}
	if event.Check.Occurrences > plugin.MAXOCCURENCE {
		fmt.Println("Event Check Ocurrences [PASS]", event.Check.Occurrences)
	} else {
		fmt.Printf("Event Check Ocurrences [FAIL] event.Occurrences:%v is less Than MaxOccurrence Configured:%v [nothing to do]", event.Check.Occurrences, plugin.MAXOCCURENCE)
		return false
	}

	return true
}

// Testeventalreadexist  teste
func Testeventalreadexist(event *types.Event) bool {

	config := httpclient.CoreClientConfig{
		URL:    server.URL,
		APIKey: "use transport layer security",
		CACert: server.Certificate(),
	}
	cl := httpclient.NewCoreClient(config)
	req := httpclient.NewEventRequest(event.Check.Namespace, event.Entity.Name, event.Check.Name)
	ev := new(corev2.Event)
	resp, err := cl.GetResource(context.Background(), req, ev)
	if err != nil {
		fmt.Println("Check Event not Found ", err)
	}
	var result bool
	switch {
	case resp.StatusCode == 404:
		result = true
	case resp.StatusCode == 200:
		result = false
	}
	return result
}


main.go
package main

import (
	"context"
	"fmt"
	"regexp"

	"github.com/sensu-community/sensu-plugin-sdk/httpclient"
	"github.com/sensu-community/sensu-plugin-sdk/sensu"
	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/sensu/sensu-go/types"
)

// Config represents the handler plugin config.
type Config struct {
	sensu.PluginConfig
	APIKey       string
	APIURL       string
	MAXOCCURENCE int64
}

var (
	plugin = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-event-keepalive-ttl-handler",
			Short:    "The Sensu Go Remove events as keepalive",
			Keyspace: "sensu.io/plugins/sensu-event-keepalive-ttl-handler/config",
		},
	}

	options = []*sensu.PluginConfigOption{
		{
			Path:      "apikey",
			Env:       "API_JEY",
			Argument:  "apikey",
			Shorthand: "k",
			Default:   "",
			Usage:     "The Sensu Api Key , use default from API_JEY env var",
			Value:     &plugin.APIKey,
		},
		{
			Path:      "url",
			Env:       "SENSU_URL",
			Argument:  "sensu_url",
			Shorthand: "u",
			Default:   "",
			Usage:     "The Sensu Api URL  , use default from SENSU_URL env var",
			Value:     &plugin.APIURL,
		},
		{
			Path:      "max_occurrence",
			Env:       "SENSU_MAX_OCCURRENCE",
			Argument:  "max_occurrence",
			Shorthand: "m",
			Default:   int64(0),
			Usage:     "The Max event Occurence after ttl to remove event   , use default from SENSU_MAX_OCCURRENCE env var",
			Value:     &plugin.MAXOCCURENCE,
		},
	}
)

func main() {
	handler := sensu.NewGoHandler(&plugin.PluginConfig, options, checkArgs, ExecuteHandler)
	handler.Execute()
}

//checkArgs check args
func checkArgs(_ *types.Event) error {
	if len(plugin.APIKey) == 0 || len(plugin.APIURL) == 0 || plugin.MAXOCCURENCE == 0 {
		return fmt.Errorf("apikey or sensu_url sensu  is empty")
	}
	return nil
}

//ExecuteHandler run eventhing
func ExecuteHandler(event *types.Event) error {

	res, err := CheckAnnotations(event)
	if err != nil {
		fmt.Println("Check Annotations error ", err)
	}
	if res {
		if CheckEventParameter(event) {
			for {
				fmt.Println("Event as correct Annotations")
				fmt.Println("Check Name :", event.Check.Name)
				fmt.Println("Namespace: ", event.Entity.Namespace)
				fmt.Println("Entity: ", event.Entity.Name)
				dl := ClientDeleteResourceEvent(event)
				if dl {
					fmt.Println("Removed Sucess")

				} else {
					fmt.Println("Fail to Remove Event by ttl")

				}
				if CheckEventAlreadexist(event) {
					break
				}
			}
			// return nil
		}
	}

	return nil
}

// CheckEventAlreadexist  Check if event alread exist
func CheckEventAlreadexist(event *types.Event) bool {

	config := httpclient.CoreClientConfig{
		URL:    plugin.APIURL,
		APIKey: plugin.APIKey,
	}
	cl := httpclient.NewCoreClient(config)
	req := httpclient.NewEventRequest(event.Check.Namespace, event.Entity.Name, event.Check.Name)
	ev := new(corev2.Event)
	resp, err := cl.GetResource(context.Background(), req, ev)
	if err != nil {
		fmt.Println("Check Event not Found ", err)
	}
	var result bool
	switch {
	case resp.StatusCode == 404:
		result = true
	case resp.StatusCode == 200:
		result = false
	}
	return result
}

// ClientDeleteResourceEvent  Remove Event from entity
func ClientDeleteResourceEvent(event *types.Event) bool {
	config := httpclient.CoreClientConfig{
		URL:    plugin.APIURL,
		APIKey: plugin.APIKey,
	}
	if CheckEventParameter(event) {

		cl := httpclient.NewCoreClient(config)
		req := httpclient.NewEventRequest(event.Check.Namespace, event.Entity.Name, event.Check.Name)
		resp, err := cl.DeleteResource(context.Background(), req)
		if err != nil {
			fmt.Println("Remove Event Check is not possible ")
			return false
		}
		switch {

		case resp.StatusCode == 204:
			fmt.Printf("HTTP Response %v\n", resp.Status)
			fmt.Printf("Event Check %v Removed Success from Entity %v \n", event.Check.Name, event.Entity.Name)
			return true

		case resp.StatusCode == 404:

			fmt.Printf("Event Check %v has already been removed from Entity %v \n", event.Check.Name, event.Entity.Name)
			return false
		}
	} else {
		fmt.Println("\nEvent not inclued on ttl parameters ")
		return false
	}
	return false
}

// CheckAnnotations Check event annotations
func CheckAnnotations(event *types.Event) (bool, error) {

	if event.Check.Annotations != nil {
		for key, value := range event.Check.Annotations {
			if key == "sensu.io/plugins/sensu-event-keepalive-ttl-handler/conf/enable" {
				if value == "true" {
					return true, nil
				}

			}

		}

	}
	return false, nil
}

//CheckEventParameter Check event
func CheckEventParameter(event *types.Event) bool {

	re, err := regexp.Compile(`\w+\s\w+\s\w+\s\w+\s\d+\s\w+\s\w+`)
	if err != nil {
		fmt.Println("Regex Compile Error ")
		return false
	}

	if re.MatchString(event.Check.Output) {
		fmt.Println("Event Check Output [PASS]", event.Check.Output)
	} else {
		fmt.Println("Event Check Output [FAIL] Event not Match with Regex ttl [nothing to do]", event.Check.Output)
		return false
	}

	if event.Check.Status == 1 {
		fmt.Println("Event Check Status [PASS]", event.Check.Status)
	} else {
		fmt.Println("Event Check Status [FAIL], Event not Match with Check Status 1 [nothing to do]", event.Check.Status)
		return false
	}
	if event.Check.Occurrences > plugin.MAXOCCURENCE {
		fmt.Println("Event Check Ocurrences [PASS]", event.Check.Occurrences)
	} else {
		fmt.Printf("Event Check Ocurrences [FAIL] event.Occurrences:%v is less Than MAXOCCURENCE Configured:%v [nothing to do]", event.Check.Occurrences, plugin.MAXOCCURENCE)
		return false
	}

	return true
}
